src/linalg.jl

```diff
@@ -182,24 +182,27 @@ end
     end
 #end
 
-@inline diagm(v::StaticVector, k::Type{Val{D}}=Val{0}) where {D} = _diagm(Size(v), v, k)
-@generated function _diagm(::Size{S}, v::StaticVector, ::Type{Val{D}}) where {S,D}
-    S1 = S[1]
-    Snew1 = S1+abs(D)
-    Snew = (Snew1, Snew1)
-    Lnew = Snew1 * Snew1
-    T = eltype(v)
-    ind = diagind(Snew1, Snew1, D)
-    exprs = fill(:(zero($T)), Lnew)
-    for n = 1:S[1]
-        exprs[ind[n]] = :(v[$n])
+@generated function diagm(kvs::Pair{<:Val,<:StaticVector}...)
+    N = maximum(abs(kv.parameters[1].parameters[1]) + length(kv.parameters[2]) for kv in kvs)
+    X = [Symbol("x_$(i)_$(j)") for i in 1:N, j in 1:N]
+    T = promote_type((eltype(kv.parameters[2]) for kv in kvs)...)
+    exprs = fill(:(zero($T)), N*N)
+    for m in eachindex(kvs)
+        kv = kvs[m]
+        ind = diagind(N, N, kv.parameters[1].parameters[1])
+        for n = 1:length(kv.parameters[2])
+            exprs[ind[n]] = :(kvs[$m].second[$n])
+        end
     end
     return quote
         $(Expr(:meta, :inline))
-        @inbounds return similar_type($v, Size($Snew))(tuple($(exprs...)))
+        @inbounds return SMatrix{$N,$N,$T}(tuple($(exprs...)))
     end
 end
 
+# old interface, to be deprecated/deleted eventually
+@inline diagm(v::StaticVector, k::Type{Val{D}}=Val{0}) where {D} = diagm(k() => v)
```

Comment:
This method should use `Val` instances on v0.7.

Reply:
Or I guess not... maybe we just add a deprecation immediately...

Reply:
Or define as is on 0.6 and as deprecated on 0.7-? I was unsure what to do here.

Reply:
Yeah I think you are right - if we simply deprecate this method on v0.7 everything should follow `Base` and `LinearAlgerbra` (and leave as-is on v0.6).